howso/engine/trainee.py

```diff
@@ -341,86 +349,79 @@ def metadata(self) -> Optional[Dict[str, Any]]:
         return deepcopy(self._metadata)
 
     @property
-    def calculated_matrices(self) -> Optional[Dict[str, DataFrame]]:
+    def calculated_matrices(self) -> None | Dict[str, DataFrame]:
         """
         The calculated matrices.
 
         Returns
         -------
-        dict
+        None | dict of str -> DataFrame
             The calculated matrices.
         """
         return self._calculated_matrices
 
     @property
-    def default_action_features(self) -> Optional[List[str]]:
+    def default_action_features(self) -> None | List[str]:
         """
         The default action features of the trainee.
 
         .. WARNING::
-            This returns a deep `copy` of the default action features. To
-            update them, use the method :func:`set_default_features`.
+            This returns a deep copy of the default action features. To
+            update them, use the method :meth:`set_default_features`.
 
         Returns
         -------
-        list of str or None
+        None | list of str
             The default action feature names for the trainee.
         """
         return deepcopy(self._default_action_features)
 
     @property
-    def default_context_features(self) -> Optional[List[str]]:
+    def default_context_features(self) -> None | List[str]:
         """
         The default context features of the trainee.
 
         .. WARNING::
-            This returns a deep `copy` of the default context features. To
-            update them, use the method :func:`set_default_features`.
+            This returns a deep copy of the default context features. To
+            update them, use the method :meth:`set_default_features`.
 
         Returns
         -------
-        list of str or None
+        None | list of str
             The default context feature names for the trainee.
         """
         return deepcopy(self._default_context_features)
 
     @property
-    def active_session(self) -> Union[Session, None]:
+    def active_session(self) -> Session | None:
         """
         The active session.
 
         Returns
         -------
-        Session
-            The session instance.
+        Session | None
+            The session instance, if it exists.
         """
         if isinstance(self.client, AbstractHowsoClient) and self.client.active_session:
             return Session.from_openapi(self.client.active_session, client=self.client)
 
-    def save(self, file_path: Optional[Union[Path, str]] = None) -> None:
+    def save(self, file_path: Optional[PathLike] = None):
         """
         Save a Trainee to disk.
 
         Parameters
         ----------
-        file_path : Path or str, optional
+        file_path : str | bytes | PathLike, optional
```

Comment:
Isn't this redundant since `PathLike` is already `str | bytes | Path`?

Reply:
Perhaps you meant for this to be `str | bytes | Path`?

Reply:
It's a bit confusing. `os.PathLike` refers to more than just `pathlib.Paths` and is in fact an ABC that backs most paths in Python. The `PathLike` in this docstring is referring to `os.PathLike` rather than the `TypeAlias` I defined at the top of the file.

Reply:
Would it be clearer if I did:

```suggestion
        file_path : str | bytes | os.PathLike, optional
```